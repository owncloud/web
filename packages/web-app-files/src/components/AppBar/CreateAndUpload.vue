<template>
  <div v-if="showActions" class="create-and-upload-actions oc-flex-inline oc-mr-s">
    <template v-if="createFileActionsAvailable">
      <span v-oc-tooltip="newButtonTooltip">
        <oc-button
          id="new-file-menu-btn"
          key="new-file-menu-btn-enabled"
          v-oc-tooltip="hideButtonLabels ? $gettext('New') : ''"
          :aria-label="newButtonAriaLabel"
          appearance="filled"
          variation="primary"
          :disabled="uploadOrFileCreationBlocked"
        >
          <oc-icon name="add" />
          <span v-if="!hideButtonLabels" v-text="$gettext('New')" />
        </oc-button>
      </span>
      <oc-drop
        drop-id="new-file-menu-drop"
        toggle="#new-file-menu-btn"
        class="oc-width-auto"
        mode="click"
        close-on-click
        padding-size="small"
      >
        <oc-list id="create-list" :class="areFileExtensionsShown ? 'expanded-list' : null">
          <li class="create-list-folder oc-menu-item-hover">
            <oc-button
              id="new-folder-btn"
              class="oc-width-1-1"
              justify-content="left"
              appearance="raw"
              @click="createNewFolderAction"
            >
              <resource-icon :resource="folderIconResource" size="medium" />
              <span v-text="$gettext('Folder')" />
            </oc-button>
          </li>
        </oc-list>
        <oc-list
          v-for="(group, groupIndex) in createFileActionsGroups"
          :key="`file-creation-group-${groupIndex}`"
        >
          <li
            v-for="(fileAction, fileActionIndex) in group"
            :key="`file-creation-item-${groupIndex}-${fileActionIndex}`"
            class="create-list-file oc-menu-item-hover"
          >
            <oc-button
              appearance="raw"
              class="oc-width-1-1"
              justify-content="left"
              :class="['new-file-btn-' + fileAction.ext]"
              @click="fileAction.handler"
            >
              <resource-icon :resource="getIconResource(fileAction)" size="medium" />
              <span class="create-list-file-item-text">{{ fileAction.label() }}</span>
              <span
                v-if="areFileExtensionsShown && fileAction.ext"
                class="create-list-file-item-extension"
              >
                {{ fileAction.ext }}
              </span>
            </oc-button>
          </li>
        </oc-list>
        <oc-list>
          <li class="create-list-shortcut oc-menu-item-hover">
            <oc-button
              id="new-shortcut-btn"
              class="oc-width-1-1"
              justify-content="left"
              appearance="raw"
              @click="createNewShortcutAction"
            >
              <oc-icon name="external-link" size="medium" />
              <span v-text="$gettext('Shortcut')" />
              <span
                v-if="areFileExtensionsShown"
                class="create-list-file-item-extension"
                v-text="'url'"
              />
            </oc-button>
          </li>
        </oc-list>
      </oc-drop>
    </template>
    <template v-else>
      <span v-oc-tooltip="newButtonTooltip">
        <oc-button
          id="new-folder-btn"
          v-oc-tooltip="hideButtonLabels ? $gettext('New Folder') : ''"
          appearance="filled"
          variation="primary"
          :aria-label="newButtonAriaLabel"
          :disabled="uploadOrFileCreationBlocked"
          @click="createNewFolderAction"
        >
          <oc-icon name="resource-type-folder" />
          <span v-if="!hideButtonLabels" v-text="$gettext('New Folder')" />
        </oc-button>
      </span>
    </template>
    <span v-oc-tooltip="uploadButtonTooltip">
      <oc-button
        id="upload-menu-btn"
        key="upload-menu-btn-enabled"
        v-oc-tooltip="hideButtonLabels ? $gettext('Upload') : ''"
        :aria-label="uploadButtonAriaLabel"
        :disabled="uploadOrFileCreationBlocked"
        appearance="outline"
        variation="passive"
      >
        <oc-icon name="upload" fill-type="line" />
        <span v-if="!hideButtonLabels" v-text="$gettext('Upload')" />
      </oc-button>
    </span>
    <oc-drop
      drop-id="upload-menu-drop"
      toggle="#upload-menu-btn"
      mode="click"
      class="oc-width-auto"
      close-on-click
      padding-size="small"
      @show-drop="showDrop"
    >
      <oc-list id="upload-list">
        <li class="oc-menu-item-hover">
          <resource-upload ref="folder-upload" btn-class="oc-width-1-1" />
        </li>
        <li class="oc-menu-item-hover">
          <resource-upload ref="file-upload" btn-class="oc-width-1-1" :is-folder="true" />
        </li>
      </oc-list>
      <oc-list v-if="extensionActions.length" id="extension-list">
        <li
          v-for="(action, key) in extensionActions"
          :key="`${key}-${actionKeySuffix}`"
          v-oc-tooltip="
            isActionDisabled(action) && action.disabledTooltip ? action.disabledTooltip() : null
          "
          class="oc-menu-item-hover"
        >
          <oc-button
            class="oc-width-1-1"
            :class="action.class"
            appearance="raw"
            justify-content="left"
            :disabled="isActionDisabled(action)"
            @click="action.handler"
          >
            <oc-icon :name="action.icon" fill-type="line" />
            <span v-text="action.label()"
          /></oc-button>
        </li>
      </oc-list>
    </oc-drop>
    <div
      v-if="showPasteHereButton"
      id="clipboard-btns"
      v-oc-tooltip="
        uploadOrFileCreationBlocked ? $gettext('You have no permission to paste files here!') : ''
      "
      class="oc-button-group"
      :class="{ disabled: uploadOrFileCreationBlocked }"
    >
      <oc-button
        :disabled="uploadOrFileCreationBlocked"
        class="paste-files-btn"
        @click="pasteFileAction"
      >
        <oc-icon fill-type="line" name="clipboard" />
        <span v-text="$gettext('Paste here')" />
      </oc-button>
      <oc-button
        :disabled="uploadOrFileCreationBlocked"
        class="clear-clipboard-btn"
        @click="clearClipboard"
      >
        <oc-icon fill-type="line" name="close" />
      </oc-button>
    </div>
  </div>
</template>

<script lang="ts">
import {
  FileAction,
  isLocationPublicActive,
  isLocationSpacesActive,
  useClipboardStore,
  useFileActions,
  useFileActionsCreateNewShortcut,
  useMessages,
  useResourcesStore,
  useSharesStore,
  useSpacesStore,
  useUserStore
} from '@ownclouders/web-pkg'
import { useActiveLocation } from '@ownclouders/web-pkg'
import {
  useFileActionsCreateNewFile,
  useFileActionsCreateNewFolder,
  useFileActionsPaste,
  useRequest,
  useClientService
} from '@ownclouders/web-pkg'

import ResourceUpload from './Upload/ResourceUpload.vue'

import {
  computed,
  defineComponent,
  onMounted,
  onBeforeUnmount,
  PropType,
  unref,
  watch,
  ref
} from 'vue'
import { eventBus } from '@ownclouders/web-pkg'
import {
  Resource,
  SpaceResource,
  isPublicSpaceResource,
  isShareSpaceResource
} from '@ownclouders/web-client'
import { useService, useUpload, UppyService, UploadResult } from '@ownclouders/web-pkg'
import { HandleUpload } from '../../HandleUpload'
import { useRoute } from 'vue-router'
import { useGettext } from 'vue3-gettext'
import { useExtensionRegistry } from '@ownclouders/web-pkg'
import { Action, ResourceIcon } from '@ownclouders/web-pkg'
import { v4 as uuidv4 } from 'uuid'
import { storeToRefs } from 'pinia'
import { uploadMenuExtensionPoint } from '../../extensionPoints'

export default defineComponent({
  components: {
    ResourceIcon,
    ResourceUpload
  },
  props: {
    space: {
      type: Object as PropType<SpaceResource>,
      required: true
    },
    item: {
      type: String,
      required: false,
      default: null
    },
    limitedScreenSpace: {
      type: Boolean,
      default: false,
      required: false
    },
    itemId: {
      type: [String, Number],
      required: false,
      default: null
    }
  },
  setup(props) {
    const uppyService = useService<UppyService>('$uppyService')
    const clientService = useClientService()
    const userStore = useUserStore()
    const spacesStore = useSpacesStore()
    const messageStore = useMessages()
    const sharesStore = useSharesStore()
    const route = useRoute()
    const language = useGettext()

    const clipboardStore = useClipboardStore()
    const { clearClipboard } = clipboardStore
    const { resources: clipboardResources } = storeToRefs(clipboardStore)

    const resourcesStore = useResourcesStore()
    const { currentFolder } = storeToRefs(resourcesStore)

    const areFileExtensionsShown = computed(() => unref(resourcesStore.areFileExtensionsShown))

    const space = computed(() => props.space)

    useUpload({ uppyService })

    if (!uppyService.getPlugin('HandleUpload')) {
      uppyService.addPlugin(HandleUpload, {
        clientService,
        language,
        route,
        space,
        userStore,
        spacesStore,
        messageStore,
        resourcesStore,
        uppyService
      })
    }

    let uploadCompletedSub: string

    const { actions: pasteFileActions } = useFileActionsPaste()
    const pasteFileAction = () => {
      return unref(pasteFileActions)[0].handler({ space: unref(space) })
    }

    const { actions: createNewFolder } = useFileActionsCreateNewFolder({ space })
    const createNewFolderAction = computed(() => unref(createNewFolder)[0].handler)

    const { actions: createNewShortcut } = useFileActionsCreateNewShortcut({ space })
    const createNewShortcutAction = computed(() => unref(createNewShortcut)[0].handler)

    const { actions: createNewFileActions } = useFileActionsCreateNewFile({ space })

    const createFileActionsGroups = computed(() => {
      const result = []
      const externalFileActions = unref(createNewFileActions).filter(({ isExternal }) => isExternal)
      if (externalFileActions.length) {
        result.push(externalFileActions)
      }
      const appFileActions = unref(createNewFileActions).filter(({ isExternal }) => !isExternal)
      if (appFileActions.length) {
        result.push(appFileActions)
      }
      return result
    })
    const createFileActionsAvailable = computed(() => {
      return unref(createFileActionsGroups).some((group) => group.length > 0)
    })

    const extensionRegistry = useExtensionRegistry()
    const extensionActions = computed(() => {
      return [
        ...extensionRegistry.requestExtensions(uploadMenuExtensionPoint).map((e) => e.action)
      ].filter((e) => e.isVisible())
    })

    const canUpload = computed(() => {
      return unref(currentFolder)?.canUpload({ user: userStore.user })
    })

    const actionKeySuffix = ref(uuidv4())
    const showDrop = () => {
      // force actions to be re-rendered when the drop is being opened
      actionKeySuffix.value = uuidv4()
    }
    const isActionDisabled = (action: Action) => {
      return action.isDisabled ? action.isDisabled() : false
    }

    const handlePasteFileEvent = (event: ClipboardEvent) => {
      // Ignore file in clipboard if there are already files from owncloud in the clipboard
      if (unref(clipboardResources).length || !unref(canUpload)) {
        return
      }
      // Browsers only allow single files to be pasted for security reasons
      const items = event.clipboardData.items
      const fileItem = [...items].find((i) => i.kind === 'file')
      if (!fileItem) {
        return
      }
      const file = fileItem.getAsFile()
      uppyService.addFiles([file])
      event.preventDefault()
    }

    const onUploadComplete = async (result: UploadResult) => {
      if (result.successful) {
        const file = result.successful[0]

        if (!file) {
          return
        }

        const { spaceId, currentFolder, currentFolderId, driveType } = file.meta
        if (!isPublicSpaceResource(unref(space))) {
          const isOwnSpace = spacesStore.spaces
            .find(({ id }) => id === spaceId)
            ?.isOwner(userStore.user)

          if (driveType === 'project' || isOwnSpace) {
            const client = clientService.graphAuthenticated
            const updatedSpace = await client.drives.getDrive(spaceId, sharesStore.graphRoles)
            spacesStore.updateSpaceField({
              id: updatedSpace.id,
              field: 'spaceQuota',
              value: updatedSpace.spaceQuota
            })
          }
        }

        const sameFolder =
          props.itemId && !isShareSpaceResource(unref(space))
            ? props.itemId.toString().startsWith(currentFolderId.toString())
            : currentFolder === props.item
        const fileIsInCurrentPath = spaceId === unref(space).id && sameFolder
        if (fileIsInCurrentPath) {
          eventBus.publish('app.files.list.load')
        }
      }
    }

    onMounted(() => {
      uploadCompletedSub = uppyService.subscribe('uploadCompleted', onUploadComplete)
      document.addEventListener('paste', handlePasteFileEvent)
    })

    onBeforeUnmount(() => {
      uppyService.removePlugin(uppyService.getPlugin('HandleUpload'))
      uppyService.unsubscribe('uploadCompleted', uploadCompletedSub)
      uppyService.removeDropTarget()
      document.removeEventListener('paste', handlePasteFileEvent)
    })

    watch(
      canUpload,
      () => {
        if (unref(canUpload)) {
          uppyService.useDropTarget({ targetSelector: '#files-view' })
        } else {
          uppyService.removeDropTarget()
        }
      },
      { immediate: true }
    )

    return {
      ...useFileActions(),
      ...useRequest(),
      clientService,
      isPublicLocation: useActiveLocation(isLocationPublicActive, 'files-public-link'),
      isSpacesGenericLocation: useActiveLocation(isLocationSpacesActive, 'files-spaces-generic'),
      canUpload,
      currentFolder,
      createNewFolder,
      createFileActionsGroups,
      createFileActionsAvailable,
      createNewFolderAction,
      createNewShortcutAction,
      extensionActions,
      pasteFileAction,
      isActionDisabled,
      actionKeySuffix,
      showDrop,
      areFileExtensionsShown,
      clearClipboard,
      clipboardResources,

      // HACK: exported for unit tests:
      onUploadComplete
    }
  },
  computed: {
    showPasteHereButton() {
      return this.clipboardResources && this.clipboardResources.length !== 0
    },
    hideButtonLabels() {
      return this.limitedScreenSpace
    },

    showActions() {
      return !(this.uploadOrFileCreationBlocked && this.isPublicLocation)
    },

    newButtonTooltip() {
      if (!this.canUpload) {
        return this.$gettext('You have no permission to create new files!')
      }
      return null
    },
    newButtonAriaLabel() {
      const tooltip = this.newButtonTooltip
      if (tooltip) {
        return tooltip
      }
      if (!this.createFileActionsAvailable) {
        return this.$gettext('New folder')
      }
      return this.$gettext('Create new files or folders')
    },
    uploadButtonTooltip() {
      if (!this.canUpload) {
        return this.$gettext('You have no permission to upload!')
      }
      return null
    },
    uploadButtonAriaLabel() {
      const tooltip = this.uploadButtonTooltip
      if (tooltip) {
        return tooltip
      }
      return this.$gettext('Upload files or folders')
    },

    uploadOrFileCreationBlocked() {
      return !this.canUpload
    },

    folderIconResource() {
      return { isFolder: true, extension: '' } as Resource
    }
  },
  methods: {
    getIconResource(fileHandler: FileAction) {
      return { type: 'file', extension: fileHandler.ext } as Resource
    }
  }
})
</script>
<style lang="scss">
#upload-menu-drop,
#new-file-menu-drop {
  min-width: 230px;

  ul:not(:first-child) {
    border-top: 1px solid var(--oc-color-border);
    padding-top: var(--oc-space-small);
  }

  ul:not(:last-child) {
    padding-bottom: var(--oc-space-small);
  }
}

.create-list-file-item-extension {
  font-weight: 400 !important;
  font-size: var(--oc-font-size-small);
  margin-left: auto;
}

.expanded-list {
  min-width: 280px !important;
}

#create-list,
#upload-list,
#new-file-menu-drop {
  .oc-icon-m svg {
    height: 100% !important;
  }
}

#clipboard-btns {
  flex-flow: inherit;

  :nth-child(1) {
    border-right: 0px !important;
    white-space: nowrap;
  }

  :nth-child(2) {
    border-left: 0px !important;
  }
}

#clipboard-btns.disabled {
  opacity: 0.6;

  button {
    opacity: 1;
  }
}

.create-and-upload-actions {
  gap: var(--oc-space-small);
}
</style>
