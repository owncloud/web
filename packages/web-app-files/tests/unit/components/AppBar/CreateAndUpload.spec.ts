import CreateAndUpload from '../../../../src/components/AppBar/CreateAndUpload.vue'
import { mock } from 'vitest-mock-extended'
import { Resource, SpaceResource } from '@ownclouders/web-client'
import {
  FileAction,
  useFileActionsCreateNewFile,
  useRequest,
  useSpacesStore,
  CapabilityStore,
  useClipboardStore,
  useFileActionsPaste,
  useExtensionRegistry,
  OcUppyFile
} from '@ownclouders/web-pkg'
import { eventBus } from '@ownclouders/web-pkg'
import { defaultPlugins, shallowMount, defaultComponentMocks } from '@ownclouders/web-test-helpers'
import { RouteLocation } from 'vue-router'
import { computed, ref } from 'vue'
import { OcButton } from '@ownclouders/design-system/components'

vi.mock('@ownclouders/web-pkg', async (importOriginal) => ({
  ...(await importOriginal<any>()),
  useRequest: vi.fn(),
  useFileActionsCreateNewFile: vi.fn(),
  useFileActions: vi.fn(),
  useFileActionsPaste: vi.fn()
}))

const elSelector = {
  component: '#create-and-upload-actions',
  newFileButton: '#new-file-menu-btn',
  uploadBtn: '#upload-menu-btn',
  resourceUpload: 'resource-upload-stub',
  newFolderBtn: '#new-folder-btn',
  clipboardBtns: '#clipboard-btns',
  pasteFilesBtn: '.paste-files-btn',
  clearClipboardBtn: '.clear-clipboard-btn'
}

describe('CreateAndUpload component', () => {
  describe('action buttons', () => {
    it('should show and be enabled if file creation is possible', () => {
      const { wrapper } = getWrapper()
      expect(
        wrapper.findComponent<typeof OcButton>(elSelector.uploadBtn).props().disabled
      ).toBeFalsy()
      expect(
        wrapper.findComponent<typeof OcButton>(elSelector.newFolderBtn).props().disabled
      ).toBeFalsy()
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('should be disabled if file creation is not possible', () => {
      const currentFolder = mock<Resource>({ canUpload: () => false })
      const { wrapper } = getWrapper({ currentFolder, createActions: [] })
      expect(
        wrapper.findComponent<typeof OcButton>(elSelector.uploadBtn).props().disabled
      ).toBeTruthy()
      expect(
        wrapper.findComponent<typeof OcButton>(elSelector.newFolderBtn).props().disabled
      ).toBeTruthy()
    })
    it('should not be visible if file creation is not possible on a public page', () => {
      const currentFolder = mock<Resource>({ canUpload: () => false })
      const { wrapper } = getWrapper({ currentFolder, currentRouteName: 'files-public-link' })
      expect(wrapper.find(elSelector.component).exists()).toBeFalsy()
    })
  })
  describe('file handlers', () => {
    it('should always show for uploading files and folders', () => {
      const { wrapper } = getWrapper()
      expect(wrapper.findAll(elSelector.resourceUpload).length).toBe(2)
    })
    it('should show entries for all new file handlers', () => {
      const { wrapper } = getWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('clipboard buttons', () => {
    it('should show if clipboard is empty', () => {
      const { wrapper } = getWrapper()
      expect(
        wrapper
          .findComponent<typeof OcButton>(`${elSelector.clipboardBtns} oc-button-stub`)
          .exists()
      ).toBeFalsy()
    })
    it('should show if clipboard is not empty', () => {
      const { wrapper } = getWrapper({ clipboardResources: [mock<Resource>()] })
      expect(wrapper.findAll(`${elSelector.clipboardBtns} .oc-button`).length).toBe(2)
    })
    it('call the "paste files"-action', async () => {
      const { wrapper, mocks } = getWrapper({
        clipboardResources: [
          mock<Resource>({
            remoteItemPath: undefined
          })
        ]
      })
      await wrapper.find(elSelector.pasteFilesBtn).trigger('click')
      expect(mocks.pasteActionHandler).toHaveBeenCalled()
    })
    it('call "clear clipboard"-action', async () => {
      const { wrapper } = getWrapper({ clipboardResources: [mock<Resource>()] })
      await wrapper.find(elSelector.clearClipboardBtn).trigger('click')
      const clipboardStore = useClipboardStore()
      expect(clipboardStore.clearClipboard).toHaveBeenCalled()
    })
  })
  describe('method "onUploadComplete"', () => {
    it.each([
      { driveType: 'personal', updated: 1 },
      { driveType: 'project', updated: 1 },
      { driveType: 'share', updated: 0 },
      { driveType: 'public', updated: 0 }
    ])('updates the space quota for supported drive types: %s', async ({ driveType, updated }) => {
      const file = mock<OcUppyFile>({ meta: { driveType, spaceId: '1' } })
      const spaces = [
        mock<SpaceResource>({ id: file.meta.spaceId, isOwner: () => driveType === 'personal' })
      ]
      const { wrapper, mocks } = getWrapper({ spaces })
      const graphMock = mocks.$clientService.graphAuthenticated
      graphMock.drives.getDrive.mockResolvedValue(mock<SpaceResource>())
      await wrapper.vm.onUploadComplete({ successful: [file], failed: [] })
      const spacesStore = useSpacesStore()
      expect(spacesStore.updateSpaceField).toHaveBeenCalledTimes(updated)
    })
    it('reloads the file list if files were uploaded to the current path', async () => {
      const eventSpy = vi.spyOn(eventBus, 'publish')
      const itemId = 'itemId'
      const space = mock<SpaceResource>({ id: '1' })
      const { wrapper, mocks } = getWrapper({ itemId, space })
      const file = mock<OcUppyFile>({
        meta: { driveType: 'project', spaceId: space.id, currentFolderId: itemId }
      })
      const graphMock = mocks.$clientService.graphAuthenticated
      graphMock.drives.getDrive.mockResolvedValue(mock<SpaceResource>())
      await wrapper.vm.onUploadComplete({ successful: [file], failed: [] })
      expect(eventSpy).toHaveBeenCalled()
    })
  })
  describe('drop target', () => {
    it('is being initialized when user can upload', () => {
      const { mocks } = getWrapper()
      expect(mocks.$uppyService.useDropTarget).toHaveBeenCalled()
    })
    it('is not being initialized when user can not upload', () => {
      const currentFolder = mock<Resource>({ canUpload: () => false })
      const { mocks } = getWrapper({ currentFolder })
      expect(mocks.$uppyService.useDropTarget).not.toHaveBeenCalled()
    })
  })
})

function getWrapper({
  clipboardResources = [],
  files = [],
  currentFolder = mock<Resource>({ canUpload: () => true }),
  currentRouteName = 'files-spaces-generic',
  space = mock<SpaceResource>(),
  spaces = [],
  itemId = undefined,
  newFileAction = false,
  areFileExtensionsShown = false,
  createActions = [
    mock<FileAction>({ label: () => 'Plain text file', ext: 'txt' }),
    mock<FileAction>({ label: () => 'Mark-down file', ext: 'md' }),
    mock<FileAction>({ label: () => 'Draw.io document', ext: 'drawio' })
  ]
}: {
  clipboardResources?: Resource[]
  files?: Resource[]
  currentFolder?: Resource
  currentRouteName?: string
  space?: SpaceResource
  spaces?: SpaceResource[]
  itemId?: string
  newFileAction?: boolean
  areFileExtensionsShown?: boolean
  createActions?: FileAction[]
} = {}) {
  const capabilities = {
    spaces: { enabled: true },
    files: { app_providers: [{ new_url: '/' }] }
  } satisfies Partial<CapabilityStore['capabilities']>

  const plugins = defaultPlugins({
    piniaOptions: {
      spacesState: { spaces },
      capabilityState: { capabilities },
      clipboardState: { resources: clipboardResources },
      resourcesStore: { areFileExtensionsShown, currentFolder, resources: files }
    }
  })

  vi.mocked(useRequest).mockImplementation(() => ({
    makeRequest: vi.fn().mockResolvedValue({ status: 200 })
  }))
  const { requestExtensions } = useExtensionRegistry()
  vi.mocked(requestExtensions).mockReturnValue([])

  const useFileActionsCreateNewFileMock = mock<ReturnType<typeof useFileActionsCreateNewFile>>()
  useFileActionsCreateNewFileMock.actions = computed(() => createActions)
  vi.mocked(useFileActionsCreateNewFile).mockReturnValue(useFileActionsCreateNewFileMock)

  const pasteActionHandler = vi.fn()
  vi.mocked(useFileActionsPaste).mockReturnValue(
    mock<ReturnType<typeof useFileActionsPaste>>({
      actions: ref([mock<FileAction>({ handler: pasteActionHandler })])
    })
  )

  const mocks = {
    ...defaultComponentMocks({ currentRoute: mock<RouteLocation>({ name: currentRouteName }) }),
    pasteActionHandler
  }

  return {
    mocks,
    wrapper: shallowMount(CreateAndUpload, {
      data: () => ({ newFileAction }),
      props: { space: space, itemId },
      global: {
        stubs: { OcButton: false },
        renderStubDefaultSlot: true,
        mocks,
        provide: mocks,
        plugins
      }
    })
  }
}
