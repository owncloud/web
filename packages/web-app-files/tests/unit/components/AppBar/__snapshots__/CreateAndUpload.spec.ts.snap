// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateAndUpload component > action buttons > should show and be enabled if file creation is possible 1`] = `
"<div class="create-and-upload-actions oc-flex-inline oc-mr-s"><span><button type="button" aria-label="Create new files or folders" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-primary oc-button-primary-filled" id="new-file-menu-btn"><!--v-if--> <!-- @slot Content of the button --> <oc-icon-stub name="add" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub> <span>New</span></button></span>
  <oc-drop-stub dropid="new-file-menu-drop" popperoptions="[object Object]" toggle="#new-file-menu-btn" position="bottom-start" mode="click" closeonclick="true" isnested="false" paddingsize="small" class="oc-width-auto">
    <oc-list-stub raw="false" id="create-list" class="">
      <li class="create-list-folder oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw oc-width-1-1" id="new-folder-btn">
          <!--v-if-->
          <!-- @slot Content of the button -->
          <resource-icon-stub resource="[object Object]" size="medium"></resource-icon-stub> <span>Folder</span>
        </button></li>
    </oc-list-stub>
    <oc-list-stub raw="false">
      <li class="create-list-file oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw oc-width-1-1 new-file-btn-txt">
          <!--v-if-->
          <!-- @slot Content of the button -->
          <resource-icon-stub resource="[object Object]" size="medium"></resource-icon-stub> <span class="create-list-file-item-text">Plain text file</span>
          <!--v-if-->
        </button></li>
      <li class="create-list-file oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw oc-width-1-1 new-file-btn-md">
          <!--v-if-->
          <!-- @slot Content of the button -->
          <resource-icon-stub resource="[object Object]" size="medium"></resource-icon-stub> <span class="create-list-file-item-text">Mark-down file</span>
          <!--v-if-->
        </button></li>
      <li class="create-list-file oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw oc-width-1-1 new-file-btn-drawio">
          <!--v-if-->
          <!-- @slot Content of the button -->
          <resource-icon-stub resource="[object Object]" size="medium"></resource-icon-stub> <span class="create-list-file-item-text">Draw.io document</span>
          <!--v-if-->
        </button></li>
    </oc-list-stub>
    <oc-list-stub raw="false">
      <li class="create-list-shortcut oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw oc-width-1-1" id="new-shortcut-btn">
          <!--v-if-->
          <!-- @slot Content of the button -->
          <oc-icon-stub name="external-link" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub> <span>Shortcut</span>
          <!--v-if-->
        </button></li>
    </oc-list-stub>
  </oc-drop-stub> <span><button type="button" aria-label="Upload files or folders" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-outline" id="upload-menu-btn"><!--v-if--> <!-- @slot Content of the button --> <oc-icon-stub name="upload" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub> <span>Upload</span></button></span>
  <oc-drop-stub dropid="upload-menu-drop" popperoptions="[object Object]" toggle="#upload-menu-btn" position="bottom-start" mode="click" closeonclick="true" isnested="false" paddingsize="small" class="oc-width-auto">
    <oc-list-stub raw="false" id="upload-list">
      <li class="oc-menu-item-hover">
        <resource-upload-stub btnlabel="" btnclass="oc-width-1-1" isfolder="false"></resource-upload-stub>
      </li>
      <li class="oc-menu-item-hover">
        <resource-upload-stub btnlabel="" btnclass="oc-width-1-1" isfolder="true"></resource-upload-stub>
      </li>
    </oc-list-stub>
    <!--v-if-->
  </oc-drop-stub>
  <!--v-if-->
</div>"
`;

exports[`CreateAndUpload component > file handlers > should show entries for all new file handlers 1`] = `
"<div class="create-and-upload-actions oc-flex-inline oc-mr-s"><span><button type="button" aria-label="Create new files or folders" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-primary oc-button-primary-filled" id="new-file-menu-btn"><!--v-if--> <!-- @slot Content of the button --> <oc-icon-stub name="add" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub> <span>New</span></button></span>
  <oc-drop-stub dropid="new-file-menu-drop" popperoptions="[object Object]" toggle="#new-file-menu-btn" position="bottom-start" mode="click" closeonclick="true" isnested="false" paddingsize="small" class="oc-width-auto">
    <oc-list-stub raw="false" id="create-list" class="">
      <li class="create-list-folder oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw oc-width-1-1" id="new-folder-btn">
          <!--v-if-->
          <!-- @slot Content of the button -->
          <resource-icon-stub resource="[object Object]" size="medium"></resource-icon-stub> <span>Folder</span>
        </button></li>
    </oc-list-stub>
    <oc-list-stub raw="false">
      <li class="create-list-file oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw oc-width-1-1 new-file-btn-txt">
          <!--v-if-->
          <!-- @slot Content of the button -->
          <resource-icon-stub resource="[object Object]" size="medium"></resource-icon-stub> <span class="create-list-file-item-text">Plain text file</span>
          <!--v-if-->
        </button></li>
      <li class="create-list-file oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw oc-width-1-1 new-file-btn-md">
          <!--v-if-->
          <!-- @slot Content of the button -->
          <resource-icon-stub resource="[object Object]" size="medium"></resource-icon-stub> <span class="create-list-file-item-text">Mark-down file</span>
          <!--v-if-->
        </button></li>
      <li class="create-list-file oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw oc-width-1-1 new-file-btn-drawio">
          <!--v-if-->
          <!-- @slot Content of the button -->
          <resource-icon-stub resource="[object Object]" size="medium"></resource-icon-stub> <span class="create-list-file-item-text">Draw.io document</span>
          <!--v-if-->
        </button></li>
    </oc-list-stub>
    <oc-list-stub raw="false">
      <li class="create-list-shortcut oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw oc-width-1-1" id="new-shortcut-btn">
          <!--v-if-->
          <!-- @slot Content of the button -->
          <oc-icon-stub name="external-link" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub> <span>Shortcut</span>
          <!--v-if-->
        </button></li>
    </oc-list-stub>
  </oc-drop-stub> <span><button type="button" aria-label="Upload files or folders" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-outline" id="upload-menu-btn"><!--v-if--> <!-- @slot Content of the button --> <oc-icon-stub name="upload" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub> <span>Upload</span></button></span>
  <oc-drop-stub dropid="upload-menu-drop" popperoptions="[object Object]" toggle="#upload-menu-btn" position="bottom-start" mode="click" closeonclick="true" isnested="false" paddingsize="small" class="oc-width-auto">
    <oc-list-stub raw="false" id="upload-list">
      <li class="oc-menu-item-hover">
        <resource-upload-stub btnlabel="" btnclass="oc-width-1-1" isfolder="false"></resource-upload-stub>
      </li>
      <li class="oc-menu-item-hover">
        <resource-upload-stub btnlabel="" btnclass="oc-width-1-1" isfolder="true"></resource-upload-stub>
      </li>
    </oc-list-stub>
    <!--v-if-->
  </oc-drop-stub>
  <!--v-if-->
</div>"
`;
